src/mm/phys_alloc.rs

```diff
@@ -0,0 +1,29 @@
+use core::alloc::{AllocError, Allocator, Layout};
+use core::ptr::{self, NonNull};
+
+use align_address::Align;
+
+use crate::arch::mm::paging::{BasePageSize, PageSize};
+
+/// Allocates whole pages of memory that correspond to a contigous range of physical frames.
+///
+/// This is useful for communicating with devices via physical memory.
+/// In this case, a struct must be contiguous in physical memory.
+pub struct PhysAlloc;
+
+unsafe impl Allocator for PhysAlloc {
+	fn allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+		assert!(layout.align() <= BasePageSize::SIZE as usize);
+		let size = layout.size().align_up(BasePageSize::SIZE as usize);
```

Comment:
Do we need to align_up, since `allocate` and `deallocate` already do?

Reply:
Yes, to return a slice with the bigger size. Once we merge `allocate` and `deallocate` into `PhysAlloc`, this will be deduplicated.